### PR TITLE
HAGO: An Automated Hardware-Aware Quantization Framework

### PR DESCRIPTION
**Pull request by ZihengJiang**
This PR implements the automated quantization framework described in the [RFC](https://discuss.tvm.apache.org/t/rfc-search-based-automated-quantization/5483). More detail can be found in the [paper draft](https://www.ziheng.org/files/hago.pdf) also. We observe that HAGO achieves speedups of 2.09x, 1.97x, and 2.48x on Intel Xeon Cascade Lake CPUs, NVIDIA Tesla T4 GPUs, ARM Cortex-A CPUs on Raspberry Pi4 relative to full precision respectively, while maintaining high post-training quantization accuracy in each case.

Still WIP.

co-authored by: @anijain2305 @hypercubestart 

## Highlights
- Hardware Awareness 
- Model-Agnostic Graph Transformation
- Search-Based Optimization

### Accuracy for the MXNet Models
image

### Accuracy for the TensorFlow Models
image

### Accuracy for the PyTorch Models
image

### Performance on x86 CPU
image

### Performance on NVIDIA GPU
image

### Performance on ARM CPU
image

### CODE REVIEW

**Comment by FrozenGene** [comment]
@ZihengJiang  As this is WIP, is it better to mark this pr as draft pr? when it is ready, we could start to review it.

**Comment by ZihengJiang** [comment]
@FrozenGene Sounds good. Thanks!
